src/cljam/util/intervals.clj

```diff
@@ -17,12 +17,55 @@
        (into (sorted-map))
        (->SortedMapIntervals)))
 
+(defn- find-nclist-overlap-intervals [^clojure.lang.Sorted nclist start end]
+  (when-let [target-intervals
+             (and nclist
+                  (->> (subseq nclist  >= start)
+                       (map second)
+                       (take-while #(<= (:start (first %)) end))))]
+    (mapcat #(cons (first %)
+                   (find-nclist-overlap-intervals (second %)
+                                                  start end))
+            target-intervals)))
+
+(deftype NclistIntervals [nclist]
+  IIntervals
+  (find-overlap-intervals* [this start end]
+    (find-nclist-overlap-intervals nclist start end)))
+
+(defn- compare-nclist [a b]
+  (let [res (Long/signum (- (:start a) (:start b)))]
+    (if (zero? res)
+      (Long/signum (- (:end b) (:end a)))
+      res)))
+(declare make-nclist*)
+
+(defn- make-nested-intervals [intervals]
+  (when (seq intervals)
+    (let [head (first intervals)
+          [nested-intervals rests]
+          (split-with #(<= (:start  head) (:start %) (:end %) (:end head))
+                      (next intervals))]
+      (cons [(:end head) [head (make-nclist* nested-intervals)]]
+            (lazy-seq (make-nested-intervals rests))))))
+
+(defn- make-nclist* [intervals]
+  (let [sorted-intervals (sort compare-nclist intervals)]
```

Comment:
This could be simplified as `(sort-by (juxt :start (comp - :end)) intervals)` if performance permits